File: backend/app/audit.py
"""ثبت تغییرات کاربران.

قاعده‌ای که در کل برنامه رعایت می‌شود: رکورد ممیزی در *همان تراکنشِ*
تغییر نوشته می‌شود. اگر تغییر برگردد، ممیزی هم برمی‌گردد؛ پس هرگز
داده و ممیزی از هم جدا نمی‌افتند.
"""
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import Any, Iterable

from sqlalchemy.orm import Session

from .models import AuditLog, User

# نام فارسی فیلدها برای نمایش در صفحه ممیزی
FIELD_FA: dict[str, str] = {
    "section": "بخش",
    "kind": "نوع سند",
    "contract_number": "شماره قرارداد",
    "party": "طرف قرارداد",
    "subject": "موضوع",
    "unit_name": "واحد مربوطه",
    "numbered_on_jalali": "تاریخ شماره‌گذاری",
    "start_jalali": "تاریخ شروع",
    "end_jalali": "تاریخ پایان",
    "duration_text": "مدت",
    "amount_rial": "مبلغ",
    "guarantees": "تضامین",
    "counterparty_people": "عوامل طرف قرارداد",
    "site": "ساختگاه/محل اجرا",
    "capacity": "ظرفیت",
    "storage_location": "محل نگهداری اصل",
    "addendum": "الحاقیه",
    "scan_link": "لینک اسکن",
    "notes": "سایر موارد",
    "role": "نقش",
    "is_active": "فعال",
    "sections": "بخش‌های مجاز",
    "full_name": "نام و نام خانوادگی",
}

ACTION_FA: dict[str, str] = {
    "login": "ورود",
    "login_failed": "ورود ناموفق",
    "logout": "خروج",
    "password_changed": "تغییر رمز عبور",
    "contract_created": "ثبت قرارداد",
    "contract_updated": "ویرایش قرارداد",
    "contract_deleted": "حذف قرارداد",
    "contract_restored": "بازیابی قرارداد",
    "attachment_added": "افزودن پیوست",
    "attachment_deleted": "حذف پیوست",
    "attachment_viewed": "مشاهده پیوست",
    "excel_imported": "درون‌ریزی اکسل",
    "excel_exported": "خروجی اکسل",
    "user_created": "ایجاد کاربر",
    "user_updated": "ویرایش کاربر",
    "access_denied": "تلاش برای دسترسی غیرمجاز",
}


@dataclass(frozen=True)
class FieldChange:
    field: str
    old: str | None
    new: str | None


def _norm(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, bool):
        return "بله" if value else "خیر"
    if isinstance(value, Decimal):
        return format(value, "f")
    if hasattr(value, "value"):  # Enum
        return str(value.value)
    text = str(value).strip()
    return text or None


def diff(before: dict[str, Any], after: dict[str, Any],
         fields: Iterable[str]) -> list[FieldChange]:
    """فقط فیلدهایی که واقعاً عوض شده‌اند برگردانده می‌شوند."""
    changes: list[FieldChange] = []
    for name in fields:
        if name not in after:
            continue
        old, new = _norm(before.get(name)), _norm(after.get(name))
        if old != new:
            changes.append(FieldChange(name, old, new))
    return changes


def record(
    db: Session,
    *,
    actor: User | None,
    action: str,
    entity: str = "",
    entity_id: str | int | None = None,
    summary: str = "",
    changes: list[FieldChange] | None = None,
    ip: str = "",
    user_agent: str = "",
) -> None:
    """یک یا چند سطر ممیزی به تراکنش جاری اضافه می‌کند (commit نمی‌کند)."""
    base = {
        "user_id": actor.id if actor else None,
        "username": actor.username if actor else "",
        "action": action,
        "entity": entity,
        "entity_id": str(entity_id) if entity_id is not None else None,
        "ip": ip[:64],
        "user_agent": user_agent[:300],
    }
    if not changes:
        db.add(AuditLog(**base, summary=summary))
        return
    for change in changes:
        db.add(AuditLog(
            **base,
            summary=summary,
            field=change.field,
            old_value=change.old,
            new_value=change.new,
        ))

File: lib/audit.js
'use strict';
/* ============================================================================
   ثبت تغییرات کاربران — زنجیره درهم‌سازی
   ---------------------------------------------------------------------------
   در بانک داده مشتریان، audit.jsonl یک فایل متنی ساده است؛ هرکس به سرور
   دسترسی داشته باشد می‌تواند سطری را پاک کند و ردی نماند.

   اینجا هر سطر، درهم‌سازی سطر پیش از خود را با خود حمل می‌کند:

       hash(n) = SHA256( hash(n-1) + محتوای سطر n )

   حذف یا تغییر هر سطر، زنجیره را از همان‌جا به بعد می‌شکند و
   `verifyChain()` دقیقاً می‌گوید از کدام سطر خراب شده است.

   این جلوی دستکاری را نمی‌گیرد — هیچ فایلی روی سرورِ در اختیارِ مهاجم امن
   نیست — اما دستکاری را از «نامرئی» به «قابل اثبات» تبدیل می‌کند. برای
   ممیزی، تشخیص‌پذیری همان چیزی است که ارزش دارد.
   ========================================================================== */
const fs     = require('fs');
const path   = require('path');
const crypto = require('crypto');
const store  = require('./store');

const FILE   = path.join(store.DATA_DIR, 'audit.jsonl');
const GENESIS = '0'.repeat(64);

/* نام فارسی فیلدها، برای نمایش در صفحه ممیزی */
const FIELD_FA = {
  section: 'بخش', kind: 'نوع سند', number: 'شماره قرارداد', party: 'طرف قرارداد',
  subject: 'موضوع', unit: 'واحد مربوطه', numberedOn: 'تاریخ شماره‌گذاری',
  start: 'تاریخ شروع', end: 'تاریخ پایان', duration: 'مدت', amount: 'مبلغ',
  guarantees: 'تضامین', people: 'عوامل طرف قرارداد', site: 'ساختگاه/محل اجرا',
  capacity: 'ظرفیت', storage: 'محل نگهداری اصل', addendum: 'الحاقیه',
  scanLink: 'لینک اسکن', notes: 'سایر موارد',
  role: 'نقش', active: 'فعال', sections: 'بخش‌های مجاز', name: 'نام و نام خانوادگی'
};

const ACTION_FA = {
  login: 'ورود', login_failed: 'ورود ناموفق', logout: 'خروج',
  password_changed: 'تغییر رمز عبور',
  contract_created: 'ثبت قرارداد', contract_updated: 'ویرایش قرارداد',
  contract_deleted: 'حذف قرارداد', contract_restored: 'بازیابی قرارداد',
  attachment_added: 'افزودن پیوست', attachment_deleted: 'حذف پیوست',
  attachment_viewed: 'مشاهده پیوست',
  excel_imported: 'درون‌ریزی اکسل', excel_exported: 'خروجی اکسل',
  user_created: 'ایجاد کاربر', user_updated: 'ویرایش کاربر',
  access_denied: 'تلاش برای دسترسی غیرمجاز'
};

const sha256 = s => crypto.createHash('sha256').update(s).digest('hex');

/* آخرین درهم‌سازی، تا برای هر نوشتن کل فایل خوانده نشود */
let lastHash = null;

function tail(){
  if (lastHash !== null) return lastHash;
  lastHash = GENESIS;
  if (fs.existsSync(FILE)){
    const lines = fs.readFileSync(FILE, 'utf8').split('\n').filter(Boolean);
    if (lines.length){
      try { lastHash = JSON.parse(lines[lines.length - 1]).h || GENESIS; } catch (_) {}
    }
  }
  return lastHash;
}

/* هر مقدار به متن قابل مقایسه تبدیل می‌شود؛ null یعنی «نبود» */
function norm(v){
  if (v === null || v === undefined) return null;
  if (typeof v === 'boolean') return v ? 'بله' : 'خیر';
  if (Array.isArray(v)) return v.length ? v.join('، ') : null;
  const t = String(v).trim();
  return t === '' ? null : t;
}

/** فقط فیلدهایی که واقعاً عوض شده‌اند برگردانده می‌شوند */
function diff(before, after, fields){
  const out = [];
  for (const f of fields){
    if (!(f in after)) continue;
    const o = norm(before[f]), n = norm(after[f]);
    if (o !== n) out.push({ field: f, old: o, new: n });
  }
  return out;
}

/**
 * یک یا چند رکورد ممیزی می‌نویسد. نوشتن هم‌زمان با تغییر داده انجام می‌شود
 * (فراخوان مسئول است پیش از پاسخ دادن، این را صدا بزند).
 */
function record({ actor, action, entity = '', entityId = null, summary = '',
                  changes = null, ip = '', agent = '' }){
  const base = {
    at: new Date().toISOString(),
    user: actor ? actor.username : '',
    userId: actor ? actor.id : null,
    action, entity,
    entityId: entityId === null ? null : String(entityId),
    summary, ip: String(ip).slice(0, 64), agent: String(agent).slice(0, 300)
  };
  const rows = (changes && changes.length)
    ? changes.map(c => Object.assign({}, base, { field: c.field, old: c.old, new: c.new }))
    : [base];

  let text = '';
  for (const row of rows){
    const body = JSON.stringify(row);
    const h = sha256(tail() + body);
    lastHash = h;
    text += JSON.stringify(Object.assign({ h }, row)) + '\n';
  }
  /* appendFileSync با پرچم 'a' اتمیک است برای نوشتن‌های کوچک‌تر از PIPE_BUF،
     و چون سرور تک‌نمونه است (قفل store) رقابتی هم در کار نیست. */
  fs.appendFileSync(FILE, text, 'utf8');
}

/** خواندن با فیلتر. جدیدترین اول. */
function query({ user = '', action = '', entityId = '', since = '', until = '',
                 limit = 200, offset = 0 } = {}){
  if (!fs.existsSync(FILE)) return { items: [], total: 0 };
  const rows = [];
  for (const line of fs.readFileSync(FILE, 'utf8').split('\n')){
    if (!line) continue;
    let r; try { r = JSON.parse(line); } catch (_) { continue; }
    if (user && r.user !== user) continue;
    if (action && r.action !== action) continue;
    if (entityId && r.entityId !== String(entityId)) continue;
    if (since && r.at < since) continue;
    if (until && r.at > until) continue;
    rows.push(r);
  }
  rows.reverse();
  return {
    total: rows.length,
    items: rows.slice(offset, offset + limit).map(r => Object.assign({}, r, {
      actionFa: ACTION_FA[r.action] || r.action,
      fieldFa: r.field ? (FIELD_FA[r.field] || r.field) : null
    }))
  };
}

/**
 * درستی زنجیره را از ابتدا بررسی می‌کند.
 * خروجی: { ok, lines, brokenAt } — brokenAt شماره اولین سطر خراب است.
 */
function verifyChain(){
  if (!fs.existsSync(FILE)) return { ok: true, lines: 0, brokenAt: null };
  const lines = fs.readFileSync(FILE, 'utf8').split('\n').filter(Boolean);
  let prev = GENESIS;
  for (let i = 0; i < lines.length; i++){
    let row; try { row = JSON.parse(lines[i]); } catch (_) { return { ok:false, lines:lines.length, brokenAt:i+1 }; }
    const { h } = row;
    const body = JSON.stringify(Object.fromEntries(Object.entries(row).filter(([k]) => k !== 'h')));
    if (h !== sha256(prev + body)) return { ok: false, lines: lines.length, brokenAt: i + 1 };
    prev = h;
  }
  return { ok: true, lines: lines.length, brokenAt: null };
}

module.exports = { record, query, diff, verifyChain, FIELD_FA, ACTION_FA, FILE };
